In [ ]:
%matplotlib inline
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import os
import wrappers
import test_utils

from test_utils import Tester, run_test_on_dataset
def make_plots(X, T, L, S, Ts, s_max=10):

    if S is not None:
        n = X.shape[1] + 1
    else:
        n = X.shape[1]
    
    plt.figure(figsize=(12, n*2.5+0.25))

    for i in range(X.shape[1]):
    
        plt.subplot(n, 1, i+1)
        ax = X[:, i]
        plt.plot(T, ax, linewidth=2, label='Original signal', color='C0')
        for t in T[L == 1]:
            plt.plot([t]*2, [ax.min(), ax.max()], color='0', linestyle='--')
        plt.ylim(ax.min(), ax.max())
        plt.xlim(0, T.max())
        plt.xticks(size=16)
        plt.yticks(size=16)
        if S is None:
            plt.xlabel("Time", size=16)
        #plt.legend(loc='upper left', fontsize=16)
        #plt.tight_layout()

    if S is not None:
        plt.subplot(n, 1, n)
        plt.plot(Ts, S, linewidth=3, label="Change-point score", color='C3')
        for t in T[L == 1]:
                plt.plot([t]*2, [-1, s_max], color='0', linestyle='--')
        plt.ylim(-1, s_max)
        plt.xlim(0, T.max())
        plt.xticks(size=16)
        plt.yticks(np.arange(0, s_max+1, 5), size=16)
        plt.xlabel("Time", size=16)
        plt.legend(loc='upper left', fontsize=16)
        #plt.tight_layout()

    #plt.show()

In [ ]:
%%time

dir_path = "../../datasets/Occupancy/"


models = [wrappers.OnlineCLF(lag_size=250, height=.1),
          wrappers.OnlineRuLSIF(lag_size=250, height=.1)]
          #wrappers.OrigRuLSIF(alpha=0.1, kernel_num=10, window_size=100, step=10, periods=1, height=0.1)]


for amodel in models:
    report = run_test_on_dataset(dir_path, amodel, downsample=1, margin=30, verbose=0, sigma=1.0,seed=12345)
    print(amodel)
    for rep in report:
        display(rep)

[  15  831 1455 2271 2895 3743 4271 5183 5759 6655 7215 8095]
[15, 831, 1455, 2271, 2895, 3743, 4271, 5183, 5759, 6655, 7215, 8095, 8143]
[15, 831, 1455, 2271, 2895, 3743, 4271, 5183, 5759, 6655, 7215, 8095, 8143] [253, 896, 1498, 2317, 2953, 3882, 4243, 5459, 6715, 7316, 8143]
[15, 831, 1455, 2271, 2895, 3743, 4271, 5183, 5759, 6655, 7215, 8095, 8143] [224, 886, 1486, 2292, 2917, 3863, 4217, 5565, 6716, 7295, 8143]
[15, 831, 1455, 2271, 2895, 3743, 4271, 5183, 5759, 6655, 7215, 8095, 8143] [221, 881, 1475, 2283, 2909, 3460, 3871, 4204, 5554, 6698, 7295, 8143]
[15, 831, 1455, 2271, 2895, 3743, 4271, 5183, 5759, 6655, 7215, 8095, 8143] [128, 882, 1470, 2290, 2904, 3426, 3875, 4212, 5492, 6681, 7279, 7870, 8143]
[15, 831, 1455, 2271, 2895, 3743, 4271, 5183, 5759, 6655, 7215, 8095, 8143] [250, 583, 1024, 4628, 6680, 6968, 8143]
[15, 831, 1455, 2271, 2895, 3743, 4271, 5183, 5759, 6655, 7215, 8095, 8143] [248, 877, 1501, 1909, 2271, 2388, 2616, 3053, 3561, 3858, 4440, 4801, 5422, 5787, 6711

,0,1,2,3
F1,0.384615,NaN,0.384615,0.384615
AUC,0.384615,NaN,0.384615,0.384615
FP,8.000000,NaN,8.000000,8.000000
DD,102.000000,NaN,102.000000,102.000000
Threshold,30.000000,NaN,30.000000,30.000000
Covering,0.768122,NaN,0.768122,0.768122
Runtime,28.363734,NaN,28.363734,28.363734


[  15  831 1455 2271 2895 3743 4271 5183 5759 6655 7215 8095]
[15, 831, 1455, 2271, 2895, 3743, 4271, 5183, 5759, 6655, 7215, 8095, 8143]
[223, 880, 1474, 2285, 2906, 3864, 4198, 5095, 5449, 6034, 6694, 7279, 8143]


,0,1,2,3
F1,0.307692,NaN,0.307692,0.307692
AUC,0.307692,NaN,0.307692,0.307692
FP,9.000000,NaN,9.000000,9.000000
DD,94.384615,NaN,94.384615,94.384615
Threshold,30.000000,NaN,30.000000,30.000000
Covering,0.787775,NaN,0.787775,0.787775
Runtime,73.749415,NaN,73.749415,73.749415


CPU times: user 316 ms, sys: 94.5 ms, total: 410 ms
Wall time: 1min 42s


In [ ]:
%%time

dir_path = "../../datasets/Occupancy/"


models = [wrappers.OrigRuLSIF(alpha=0.1, kernel_num=10, window_size=250, step=10, periods=1, height=0.1)]


for amodel in models:
    report = run_test_on_dataset(dir_path, amodel, downsample=1, margin=30, verbose=0, sigma=1.0,seed=12345)
    print(amodel)
    for rep in report:
        display(rep)

[  15  831 1455 2271 2895 3743 4271 5183 5759 6655 7215 8095]
[15, 831, 1455, 2271, 2895, 3743, 4271, 5183, 5759, 6655, 7215, 8095, 8143]
[270, 867, 1457, 2260, 2900, 3837, 4157, 5367, 5687, 6677, 6980, 7287, 8143]


,0,1,2,3
F1,0.384615,NaN,0.384615,0.384615
AUC,0.384615,NaN,0.384615,0.384615
FP,8.000000,NaN,8.000000,8.000000
DD,84.769231,NaN,84.769231,84.769231
Threshold,30.000000,NaN,30.000000,30.000000
Covering,0.798814,NaN,0.798814,0.798814
Runtime,12.170111,NaN,12.170111,12.170111


CPU times: user 832 ms, sys: 91 ms, total: 923 ms
Wall time: 12.2 s


In [ ]:
import zipfile
import json

zip_file_path = '../../datasets/MNISTSeq.zip'

with zipfile.ZipFile(zip_file_path, 'r') as zip_ref:
    zip_contents = zip_ref.namelist()
    json_file_name = zip_contents[0]
    with zip_ref.open(json_file_name) as json_file:
        data = json.load(json_file)

In [11]:
for sample in data.keys():
    df =pd.DataFrame(data[str(sample)]['data'],columns=['X'+str(i) for i in range(1,785)])
    df.insert(0,'Time',np.arange(df.shape[0]))
    df.to_csv(f'../data/MNIST_SWCPD/sample{sample}.csv',index=False)
    np.save(f"../data/MNIST_SWCPD/sample{sample}.npy",np.array(data[str(sample)]['target']))


In [5]:
%%time

dir_path = "../data/MNIST_SWCPD/"


models = [wrappers.OnlineCLF(lag_size=100, height=.1),
          wrappers.OnlineRuLSIF(lag_size=100, height=.1)]
          #wrappers.OrigRuLSIF(alpha=0.1, kernel_num=10, window_size=100, step=10, periods=1, height=0.1)]


for amodel in models:
    report = run_test_on_dataset(dir_path, amodel, downsample=1, margin=20, verbose=0, sigma=5.0,seed=12345)
    print(amodel)
    for rep in report:
        display(rep)

[200 400]
[200, 400, 600]
[200, 400, 600] [211, 363, 600]
[200, 400, 600] [198, 397, 600]
[200, 400, 600] [203, 443, 600]
[200, 400, 600] [203, 345, 396, 600]
[200, 400, 600] [600]
[200, 400, 600] [600]
[200, 400, 600] [293, 423, 600]
[200, 400, 600] [93, 162, 243, 380, 600]
[200, 400, 600] [224, 395, 600]
[200, 400, 600] [208, 448, 600]
[200, 400, 600] [220, 600]
[200, 400, 600] [223, 406, 600]
[200, 400, 600] [600]
[200, 400, 600] [96, 600]
[200, 400, 600] [216, 420, 600]
[200, 400, 600] [255, 392, 600]
[198, 397, 600]
[200 400]
[200, 400, 600]
[200, 400, 600] [215, 356, 600]
[200, 400, 600] [260, 486, 600]
[200, 400, 600] [236, 364, 600]
[200, 400, 600] [126, 320, 485, 600]
[200, 400, 600] [172, 600]
[200, 400, 600] [600]
[200, 400, 600] [88, 218, 393, 486, 600]
[200, 400, 600] [35, 181, 483, 600]
[200, 400, 600] [213, 385, 600]
[200, 400, 600] [220, 600]
[200, 400, 600] [231, 600]
[200, 400, 600] [197, 600]
[200, 400, 600] [600]
[200, 400, 600] [600]
[200, 400, 600] [86, 228, 600]


,0,1,2,3
F1,0.935926,0.097564,0.750000,1.000000
AUC,0.936667,0.097223,0.750000,1.000000
FP,0.333333,0.487950,0.000000,1.000000
DD,6.490000,3.784795,1.666667,16.400000
Threshold,20.000000,0.000000,20.000000,20.000000
Covering,0.941052,0.027837,0.893898,0.983490
Runtime,3.427849,0.765409,2.332376,4.297882


[200 400]
[200, 400, 600]
[204, 416, 600]
[200 400]
[200, 400, 600]
[210, 397, 600]
[200 400 600 800]
[200, 400, 600, 800, 1000]
[208, 408, 594, 802, 1000]
[200 400 600 800]
[200, 400, 600, 800, 1000]
[208, 412, 586, 794, 1000]
[200 400 600 800]
[200, 400, 600, 800, 1000]
[208, 429, 577, 803, 1000]
[200 400 600 800]
[200, 400, 600, 800, 1000]
[205, 368, 588, 803, 1000]
[200 400 600 800]
[200, 400, 600, 800, 1000]
[201, 401, 583, 788, 1000]
[200 400]
[200, 400, 600]
[218, 398, 600]
[200 400]
[200, 400, 600]
[206, 391, 600]
[200 400]
[200, 400, 600]
[208, 407, 600]
[200 400 600]
[200, 400, 600, 800]
[205, 393, 593, 800]
[200 400 600]
[200, 400, 600, 800]
[204, 399, 593, 800]
[200 400 600]
[200, 400, 600, 800]
[199, 424, 596, 800]
[200 400 600]
[200, 400, 600, 800]
[199, 390, 590, 800]
[200 400 600]
[200, 400, 600, 800]
[261, 428, 598, 800]


,0,1,2,3
F1,0.910000,0.167118,0.500000,1.000000
AUC,0.910000,0.167118,0.500000,1.000000
FP,0.400000,0.736788,0.000000,2.000000
DD,7.511111,4.877535,3.000000,22.750000
Threshold,20.000000,0.000000,20.000000,20.000000
Covering,0.929491,0.041308,0.804008,0.970444
Runtime,8.109012,2.078864,5.479291,11.301667


CPU times: user 2.31 s, sys: 557 ms, total: 2.86 s
Wall time: 2min 55s


In [13]:
%%time

dir_path = "../data/MNIST_SWCPD/"


models = [wrappers.OrigRuLSIF(alpha=0.1, kernel_num=10, window_size=100, step=10, periods=1, height=0.1)]


for amodel in models:
    report = run_test_on_dataset(dir_path, amodel, downsample=1, margin=20, verbose=0, sigma=5.0,seed=12345)
    print(amodel)
    for rep in report:
        display(rep)

[200 400]
[200, 400, 600]
[600]
[200 400]
[200, 400, 600]
[600]
[200 400 600 800]
[200, 400, 600, 800, 1000]
[1000]
[200 400 600 800]
[200, 400, 600, 800, 1000]
[1000]
[200 400 600 800]
[200, 400, 600, 800, 1000]
[1000]
[200 400 600 800]
[200, 400, 600, 800, 1000]
[1000]
[200 400 600 800]
[200, 400, 600, 800, 1000]
[1000]
[200 400]
[200, 400, 600]
[600]
[200 400]
[200, 400, 600]
[600]
[200 400]
[200, 400, 600]
[600]
[200 400 600]
[200, 400, 600, 800]
[800]
[200 400 600]
[200, 400, 600, 800]
[800]
[200 400 600]
[200, 400, 600, 800]
[800]
[200 400 600]
[200, 400, 600, 800]
[800]
[200 400 600]
[200, 400, 600, 800]
[800]


,0,1,2,3
F1,0.411111,0.070897,0.333333,0.500000
AUC,0.630556,0.028464,0.600000,0.666667
FP,0.000000,0.000000,0.000000,0.000000
DD,0.000000,0.000000,0.000000,0.000000
Threshold,20.000000,0.000000,20.000000,20.000000
Covering,0.261111,0.056928,0.200000,0.333333
Runtime,0.606861,0.152665,0.396035,0.873684


CPU times: user 3.13 s, sys: 2.38 s, total: 5.51 s
Wall time: 10.6 s


In [ ]:


X_train = np.loadtxt("../../datasets/HAR_NEW/UCI HAR Dataset/train/X_train.txt"   
)
y_train = np.loadtxt("../../datasets/HAR_NEW/UCI HAR Dataset/train/y_train.txt")

human = np.loadtxt("../../datasets/HAR_NEW/UCI HAR Dataset/train/subject_train.txt")


X_test = np.loadtxt("../../datasets/HAR_NEW/UCI HAR Dataset/test/X_test.txt")

y_test = np.loadtxt("../../datasets/HAR_NEW/UCI HAR Dataset/test/y_test.txt")

human_test = np.loadtxt("../../datasets/HAR_NEW/UCI HAR Dataset/test/subject_test.txt")


df_train = pd.DataFrame(X_train)
df_train['state']  =y_train
df_train['human'] = human.astype(int)

df_test = pd.DataFrame(X_test)
df_test['state']  =y_test
df_test['human'] = human_test.astype(int)

for human in df_train['human'].unique():
    tmp = df_train[df_train['human']==human]
    tmp = tmp.reset_index(drop=True)
    
    GroundTruth = tmp.index[tmp.iloc[:,-2].diff()!=0][1:]+1

    data = tmp.iloc[:,:-2].values.astype(np.float32)
    data = (data-np.mean(data))/np.std(data)


    np.save(f'../data/HAR_NEW/{human}.npy',GroundTruth.values)




for human in df_test['human'].unique():
    tmp = df_test[df_test['human']==human]
    tmp = tmp.reset_index(drop=True)
    
    GroundTruth = tmp.index[tmp.iloc[:,-2].diff()!=0][1:]+1

    data = tmp.iloc[:,:-2].values.astype(np.float32)
    data = (data-np.mean(data))/np.std(data)


    #np.savetxt(f'data{human}.csv',data,delimiter=",")
    np.save(f'../data/HAR_NEW/{human}.npy',GroundTruth.values)


In [ ]:
%%time


dir_path = "../data/HAR_NEW/"




models = [wrappers.OnlineCLF(lag_size=20, height=.1),
          wrappers.OnlineRuLSIF(lag_size=20, height=.1)]
          #wrappers.OrigRuLSIF(alpha=0.1, kernel_num=10, window_size=100, step=10, periods=1, height=0.1)]


for amodel in models:
    report = run_test_on_dataset(dir_path, amodel, downsample=1, margin=10, verbose=0, sigma=1.0,seed=12345)
    print(amodel)
    for rep in report:
        display(rep)

[ 28  52  79 126 151 177 203 214 226 238 249 297 304 331]
[28, 52, 79, 126, 151, 177, 203, 214, 226, 238, 249, 297, 304, 331, 347]
[28, 52, 79, 126, 151, 177, 203, 214, 226, 238, 249, 297, 304, 331, 347] [78, 176, 250, 347]
[28, 52, 79, 126, 151, 177, 203, 214, 226, 238, 249, 297, 304, 331, 347] [5, 23, 40, 78, 175, 209, 249, 305, 347]
[28, 52, 79, 126, 151, 177, 203, 214, 226, 238, 249, 297, 304, 331, 347] [17, 79, 125, 151, 176, 202, 248, 301, 347]
[28, 52, 79, 126, 151, 177, 203, 214, 226, 238, 249, 297, 304, 331, 347] [17, 63, 89, 135, 175, 204, 248, 301, 347]
[28, 52, 79, 126, 151, 177, 203, 214, 226, 238, 249, 297, 304, 331, 347] [70, 90, 347]
[28, 52, 79, 126, 151, 177, 203, 214, 226, 238, 249, 297, 304, 331, 347] [29, 117, 321, 347]
[28, 52, 79, 126, 151, 177, 203, 214, 226, 238, 249, 297, 304, 331, 347] [81, 100, 138, 200, 347]
[28, 52, 79, 126, 151, 177, 203, 214, 226, 238, 249, 297, 304, 331, 347] [30, 62, 92, 118, 151, 179, 251, 347]
[28, 52, 79, 126, 151, 177, 203, 214, 22

,0,1,2,3
F1,0.771795,0.096903,0.583333,0.956522
AUC,0.795499,0.084663,0.619048,0.958333
FP,0.766667,0.858360,0.000000,3.000000
DD,3.494973,1.709302,1.444444,9.500000
Threshold,10.000000,0.000000,10.000000,10.000000
Covering,0.650811,0.070501,0.490539,0.805065
Runtime,1.496997,0.310618,1.177362,2.904778


[ 28  52  79 126 151 177 203 214 226 238 249 297 304 331]
[28, 52, 79, 126, 151, 177, 203, 214, 226, 238, 249, 297, 304, 331, 347]
[22, 78, 121, 152, 175, 208, 248, 303, 323, 347]
[ 24  48  79 106 125 148 169 199 226 252 271]
[24, 48, 79, 106, 125, 148, 169, 199, 226, 252, 271, 294]
[23, 77, 114, 143, 178, 224, 270, 294]
[ 22  48  78 107 129 155 181 208 235 265 289]
[22, 48, 78, 107, 129, 155, 181, 208, 235, 265, 289, 316]
[19, 50, 78, 104, 155, 180, 235, 265, 291, 316]
[ 28  56  83 110 135 162 196 219 252 275 296]
[28, 56, 83, 110, 135, 162, 196, 219, 252, 275, 296, 320]
[21, 54, 82, 107, 162, 222, 251, 268, 320]
[ 28  52  85 115 142 173 203 228 257 284 304]
[28, 52, 85, 115, 142, 173, 203, 228, 257, 284, 304, 327]
[23, 54, 85, 116, 147, 177, 227, 264, 282, 327]
[ 31  57  82 112 136 163 193 221 247 276 297]
[31, 57, 82, 112, 136, 163, 193, 221, 247, 276, 297, 323]
[22, 33, 45, 82, 111, 163, 218, 248, 276, 323]
[ 28  55  89 117 138 162 188 220 258 284 305]
[28, 55, 89, 117, 138, 162, 1

,0,1,2,3
F1,0.790937,0.091066,0.615385,0.956522
AUC,0.813966,0.077474,0.678571,0.958333
FP,0.600000,0.723974,0.000000,2.000000
DD,3.175259,1.093156,1.100000,6.454545
Threshold,10.000000,0.000000,10.000000,10.000000
Covering,0.669925,0.078764,0.534278,0.840636
Runtime,3.599403,0.619468,2.710591,5.389486


CPU times: user 2.77 s, sys: 615 ms, total: 3.39 s
Wall time: 2min 35s


In [14]:
%%time

dir_path = "../data/HAR_NEW/"


models = [wrappers.OrigRuLSIF(alpha=0.1, kernel_num=10, window_size=20, step=10, periods=1, height=0.1)]


for amodel in models:
    report = run_test_on_dataset(dir_path, amodel, downsample=1, margin=10, verbose=0, sigma=1.0,seed=12345)
    print(amodel)
    for rep in report:
        display(rep)

[ 28  52  79 126 151 177 203 214 226 238 249 297 304 331]
[28, 52, 79, 126, 151, 177, 203, 214, 226, 238, 249, 297, 304, 331, 347]
[13, 43, 63, 83, 113, 143, 173, 203, 243, 293, 347]
[ 24  48  79 106 125 148 169 199 226 252 271]
[24, 48, 79, 106, 125, 148, 169, 199, 226, 252, 271, 294]
[13, 73, 93, 113, 133, 163, 193, 223, 243, 294]
[ 22  48  78 107 129 155 181 208 235 265 289]
[22, 48, 78, 107, 129, 155, 181, 208, 235, 265, 289, 316]
[13, 43, 73, 103, 123, 153, 183, 203, 223, 253, 273, 316]
[ 28  56  83 110 135 162 196 219 252 275 296]
[28, 56, 83, 110, 135, 162, 196, 219, 252, 275, 296, 320]
[23, 53, 73, 103, 123, 153, 183, 213, 243, 283, 320]
[ 28  52  85 115 142 173 203 228 257 284 304]
[28, 52, 85, 115, 142, 173, 203, 228, 257, 284, 304, 327]
[23, 43, 73, 133, 163, 193, 223, 243, 273, 327]
[ 31  57  82 112 136 163 193 221 247 276 297]
[31, 57, 82, 112, 136, 163, 193, 221, 247, 276, 297, 323]
[23, 53, 73, 103, 133, 153, 183, 213, 243, 273, 323]
[ 28  55  89 117 138 162 188 220 258 

,0,1,2,3
F1,0.703356,0.105488,0.518519,0.956522
AUC,0.718689,0.099320,0.536932,0.958333
FP,2.166667,1.147211,0.000000,4.000000
DD,7.015580,1.077376,4.818182,9.090909
Threshold,10.000000,0.000000,10.000000,10.000000
Covering,0.543135,0.060513,0.416439,0.652040
Runtime,0.105930,0.012183,0.087316,0.129055


CPU times: user 1.48 s, sys: 152 ms, total: 1.63 s
Wall time: 5.06 s


In [ ]:
%%time

dir_path = "../../datasets/has2023DataR/"


models = [wrappers.OnlineCLF(lag_size=250, height=.1),
          wrappers.OnlineRuLSIF(lag_size=250, height=.1)]
          #wrappers.OrigRuLSIF(alpha=0.1, kernel_num=10, window_size=100, step=10, periods=1, height=0.1)]


for amodel in models:
    report = run_test_on_dataset(dir_path, amodel, downsample=1, margin=100, verbose=0, sigma=1.0,seed=12345)
    print(amodel)
    for rep in report:
        display(rep)

[1275 2746 4184]
[1275, 2746, 4184, 5597]
[1275, 2746, 4184, 5597] [1297, 1625, 1777, 2317, 2772, 4273, 4681, 5314, 5597]
[1275, 2746, 4184, 5597] [143, 967, 1293, 1610, 1753, 1934, 2320, 2768, 4284, 4657, 5016, 5314, 5597]
[1275, 2746, 4184, 5597] [220, 1290, 1770, 2322, 2758, 3330, 4284, 4702, 5034, 5314, 5597]
[1275, 2746, 4184, 5597] [238, 594, 750, 1279, 1783, 2289, 2733, 3136, 3330, 3547, 4268, 4653, 5029, 5314, 5597]
[1275, 2746, 4184, 5597] [844, 1395, 2864, 4330, 4809, 5597]
[1275, 2746, 4184, 5597] [421, 868, 1326, 1863, 2758, 4289, 4810, 5314, 5597]
[1275, 2746, 4184, 5597] [421, 864, 1299, 2758, 3538, 4301, 4772, 5287, 5597]
[1275, 2746, 4184, 5597] [436, 965, 1287, 2205, 2768, 3363, 4206, 5195, 5597]
[1275, 2746, 4184, 5597] [145, 1296, 1769, 2270, 2766, 4267, 4669, 5305, 5597]
[1275, 2746, 4184, 5597] [709, 1290, 1944, 2323, 2759, 3345, 3794, 4277, 4638, 5035, 5305, 5597]
[1275, 2746, 4184, 5597] [298, 1286, 1788, 2313, 2746, 4271, 4651, 5042, 5305, 5597]
[1275, 2746, 418

,0,1,2,3
F1,0.809650,0.180477,0.400000,1.000000
AUC,0.852137,0.125307,0.625000,1.000000
FP,1.272727,1.202451,0.000000,4.000000
DD,454.422078,874.256752,0.000000,4006.000000
Threshold,100.000000,0.000000,100.000000,100.000000
Covering,0.860011,0.125635,0.521672,1.000000
Runtime,16.296229,10.239380,0.732900,38.045784


[1275 2746 4184]
[1275, 2746, 4184, 5597]
[1319, 2804, 3260, 4307, 5597]
[]
[954]
[262, 954]
[1513 1935]
[1513, 1935, 3423]
[68, 1628, 2011, 3423]
[1429 2844 4098]
[1429, 2844, 4098, 5952]
[1485, 2886, 3290, 4177, 5952]
[]
[6200]
[370, 996, 6200]
[4175 5290 6760 7750]
[4175, 5290, 6760, 7750, 8740]
[229, 361, 917, 4189, 5350, 5925, 6762, 7222, 7770, 8740]
[1525 3070 4320 5690 6930]
[1525, 3070, 4320, 5690, 6930, 8260]
[937, 1546, 3091, 4396, 5729, 6977, 8260]
[1473 2820]
[1473, 2820, 4296]
[105, 912, 1517, 2879, 4296]
[1575 3125]
[1575, 3125, 4504]
[304, 1598, 3152, 4504]
[1290]
[1290, 2860]
[1312, 2860]
[1291 2884 4272 5977]
[1291, 2884, 4272, 5977, 7238]
[285, 1408, 2912, 4309, 6009, 6494, 7238]
[1443 1857]
[1443, 1857, 2781]
[1471, 1909, 2781]
[1611 2979]
[1611, 2979, 4111]
[1628, 3016, 4111]
[]
[1303]
[304, 1303]
[1440 3006 4383 4862 6381]
[1440, 3006, 4383, 4862, 6381, 7948]
[128, 1460, 3023, 4434, 6401, 7948]
[1542 3072 4590 6057 7485 8976]
[1542, 3072, 4590, 6057, 7485, 8976, 10

,0,1,2,3
F1,0.764838,0.170484,0.500000,1.000000
AUC,0.812013,0.128957,0.583333,1.000000
FP,1.500000,1.300183,0.000000,5.000000
DD,496.623972,800.154949,0.000000,3678.000000
Threshold,100.000000,0.000000,100.000000,100.000000
Covering,0.839439,0.109913,0.545364,1.000000
Runtime,42.929937,25.708809,2.192463,95.742753


CPU times: user 4.19 s, sys: 1.48 s, total: 5.68 s
Wall time: 21min 44s


In [ ]:
%%time

dir_path = "../../datasets/has2023DataR/"


models = [wrappers.OrigRuLSIF(alpha=0.1, kernel_num=10, window_size=250, step=10, periods=1, height=0.1)]


for amodel in models:
    report = run_test_on_dataset(dir_path, amodel, downsample=1, margin=100, verbose=0, sigma=1.0,seed=12345)
    print(amodel)
    for rep in report:
        display(rep)

[1275 2746 4184]
[1275, 2746, 4184, 5597]
[1267, 1737, 2740, 4200, 4757, 5270, 5597]
[]
[954]
[367, 954]
[1513 1935]
[1513, 1935, 3423]
[1530, 3117, 3423]
[1429 2844 4098]
[1429, 2844, 4098, 5952]
[1417, 2817, 4110, 5952]
[]
[6200]
[270, 540, 747, 1280, 1800, 2187, 2587, 3527, 3937, 6200]
[4175 5290 6760 7750]
[4175, 5290, 6760, 7750, 8740]
[270, 877, 4410, 5270, 5717, 6757, 7040, 7197, 7740, 8740]
[1525 3070 4320 5690 6930]
[1525, 3070, 4320, 5690, 6930, 8260]
[840, 1517, 3047, 4340, 5670, 6917, 8260]
[1473 2820]
[1473, 2820, 4296]
[1470, 2827, 3247, 4296]
[1575 3125]
[1575, 3125, 4504]
[320, 1557, 2007, 3107, 4504]
[1290]
[1290, 2860]
[1287, 2860]
[1291 2884 4272 5977]
[1291, 2884, 4272, 5977, 7238]
[317, 590, 910, 1270, 1750, 2877, 4260, 5977, 6490, 7238]
[1443 1857]
[1443, 1857, 2781]
[1437, 1870, 2781]
[1611 2979]
[1611, 2979, 4111]
[1610, 2970, 4111]
[]
[1303]
[467, 830, 980, 1303]
[1440 3006 4383 4862 6381]
[1440, 3006, 4383, 4862, 6381, 7948]
[1420, 2997, 4420, 4600, 4870, 6357

ValueError: cannot reshape array of size 0 into shape (0,newaxis)

In [15]:
s = "../data/MNIST_SWCPD/sample0.csv"

target = s.split("/")[-1].split(".")[0]+".npy"

s.split("/")[:-1]

['..', 'data', 'MNIST_SWCPD']

# MNIST

In [2]:
%%time

dir_path = "../data/mnist/"

# models = [wrappers.RupturesBinseg(model='rbf', jump=10), 
#           wrappers.RupturesPelt(model='rbf', jump=10), 
#           wrappers.RupturesWindow(model='rbf', jump=10, width=100), 
#           wrappers.OnlineCLF(lag_size=100, height=.1), 
#           wrappers.OnlineRuLSIF(lag_size=100, height=.1), 
#           wrappers.OrigRuLSIF(alpha=0.1, kernel_num=10, window_size=100, step=10, periods=1, height=0.1)]

models = [wrappers.OnlineCLF(lag_size=100, height=.1), 
          wrappers.OnlineRuLSIF(lag_size=100, height=.1),
          wrappers.OrigRuLSIF(alpha=0.1, kernel_num=10, window_size=100, step=10, periods=1, height=0.1)]


for amodel in models:
    report = run_test_on_dataset(dir_path, amodel, downsample=1, margin=20, verbose=0, sigma=5.0)
    print(amodel)
    for rep in report:
        display(rep)

True
True
True
True
True
True
True
True
True
True


,0,1,2,3
F1,0.666667,0.134687,0.476190,0.857143
AUC,0.668182,0.134993,0.477273,0.859091
FP,4.000000,1.414214,2.000000,6.000000
DD,10.950000,3.479863,5.400000,16.000000
Threshold,10.000000,0.000000,10.000000,10.000000
Covering,0.888061,0.033462,0.839725,0.941963
Runtime,7.640872,3.268985,6.488677,16.942395


,0,1,2,3
F1,0.771429,0.104810,0.571429,0.952381
AUC,0.773182,0.105048,0.572727,0.954545
FP,2.900000,1.100505,1.000000,5.000000
DD,10.950000,3.479863,5.400000,16.000000
Threshold,20.000000,0.000000,20.000000,20.000000
Covering,0.888061,0.033462,0.839725,0.941963
Runtime,7.640872,3.268985,6.488677,16.942395


,0,1,2,3
F1,0.857143,0.077762,0.761905,0.952381
AUC,0.859091,0.077938,0.763636,0.954545
FP,2.000000,0.816497,1.000000,3.000000
DD,10.950000,3.479863,5.400000,16.000000
Threshold,30.000000,0.000000,30.000000,30.000000
Covering,0.888061,0.033462,0.839725,0.941963
Runtime,7.640872,3.268985,6.488677,16.942395


,0,1,2,3
F1,0.895238,0.080312,0.761905,0.952381
AUC,0.897273,0.080494,0.763636,0.954545
FP,1.600000,0.843274,1.000000,3.000000
DD,10.950000,3.479863,5.400000,16.000000
Threshold,40.000000,0.000000,40.000000,40.000000
Covering,0.888061,0.033462,0.839725,0.941963
Runtime,7.640872,3.268985,6.488677,16.942395


,0,1,2,3
F1,0.933333,0.040156,0.857143,0.952381
AUC,0.935455,0.040247,0.859091,0.954545
FP,1.200000,0.421637,1.000000,2.000000
DD,10.950000,3.479863,5.400000,16.000000
Threshold,50.000000,0.000000,50.000000,50.000000
Covering,0.888061,0.033462,0.839725,0.941963
Runtime,7.640872,3.268985,6.488677,16.942395


True
True
True
True
True
True
True
True
True
True


,0,1,2,3
F1,0.605195,0.145633,0.272727,0.761905
AUC,0.607121,0.145642,0.275000,0.763636
FP,4.800000,1.751190,3.000000,9.000000
DD,11.503636,3.832362,7.500000,20.636364
Threshold,10.000000,0.000000,10.000000,10.000000
Covering,0.886684,0.029264,0.818179,0.919894
Runtime,17.047756,0.215549,16.734643,17.408990


,0,1,2,3
F1,0.802165,0.068783,0.727273,0.952381
AUC,0.804924,0.068562,0.733333,0.954545
FP,2.700000,0.823273,1.000000,4.000000
DD,11.503636,3.832362,7.500000,20.636364
Threshold,20.000000,0.000000,20.000000,20.000000
Covering,0.886684,0.029264,0.818179,0.919894
Runtime,17.047756,0.215549,16.734643,17.408990


,0,1,2,3
F1,0.906061,0.053374,0.818182,0.952381
AUC,0.909167,0.052662,0.825000,0.954545
FP,1.600000,0.699206,1.000000,3.000000
DD,11.503636,3.832362,7.500000,20.636364
Threshold,30.000000,0.000000,30.000000,30.000000
Covering,0.886684,0.029264,0.818179,0.919894
Runtime,17.047756,0.215549,16.734643,17.408990


,0,1,2,3
F1,0.925108,0.048894,0.818182,0.952381
AUC,0.928258,0.047630,0.825000,0.954545
FP,1.400000,0.699206,1.000000,3.000000
DD,11.503636,3.832362,7.500000,20.636364
Threshold,40.000000,0.000000,40.000000,40.000000
Covering,0.886684,0.029264,0.818179,0.919894
Runtime,17.047756,0.215549,16.734643,17.408990


,0,1,2,3
F1,0.934199,0.032511,0.857143,0.952381
AUC,0.937424,0.031710,0.859091,0.954545
FP,1.300000,0.483046,1.000000,2.000000
DD,11.503636,3.832362,7.500000,20.636364
Threshold,50.000000,0.000000,50.000000,50.000000
Covering,0.886684,0.029264,0.818179,0.919894
Runtime,17.047756,0.215549,16.734643,17.408990


True
True
True
True
True
True
True
True
True
True


,0,1,2,3
F1,0.503578,0.096282,0.333333,0.666667
AUC,0.512599,0.099389,0.337500,0.675000
FP,3.400000,1.173788,2.000000,5.000000
DD,9.957738,3.029291,5.500000,15.375000
Threshold,10.000000,0.000000,10.000000,10.000000
Covering,0.629847,0.072006,0.531934,0.726995
Runtime,0.755754,0.019989,0.742016,0.803356


,0,1,2,3
F1,0.691572,0.070669,0.588235,0.777778
AUC,0.703254,0.068458,0.607143,0.787500
FP,1.700000,0.674949,1.000000,3.000000
DD,9.957738,3.029291,5.500000,15.375000
Threshold,20.000000,0.000000,20.000000,20.000000
Covering,0.629847,0.072006,0.531934,0.726995
Runtime,0.755754,0.019989,0.742016,0.803356


,0,1,2,3
F1,0.735501,0.067535,0.588235,0.842105
AUC,0.747758,0.062469,0.607143,0.844444
FP,1.300000,0.483046,1.000000,2.000000
DD,9.957738,3.029291,5.500000,15.375000
Threshold,30.000000,0.000000,30.000000,30.000000
Covering,0.629847,0.072006,0.531934,0.726995
Runtime,0.755754,0.019989,0.742016,0.803356


,0,1,2,3
F1,0.746027,0.075501,0.588235,0.842105
AUC,0.758313,0.069343,0.607143,0.844444
FP,1.200000,0.421637,1.000000,2.000000
DD,9.957738,3.029291,5.500000,15.375000
Threshold,40.000000,0.000000,40.000000,40.000000
Covering,0.629847,0.072006,0.531934,0.726995
Runtime,0.755754,0.019989,0.742016,0.803356


,0,1,2,3
F1,0.757792,0.054399,0.705882,0.842105
AUC,0.770456,0.046944,0.728571,0.844444
FP,1.100000,0.316228,1.000000,2.000000
DD,9.957738,3.029291,5.500000,15.375000
Threshold,50.000000,0.000000,50.000000,50.000000
Covering,0.629847,0.072006,0.531934,0.726995
Runtime,0.755754,0.019989,0.742016,0.803356


CPU times: user 2.56 s, sys: 1.02 s, total: 3.58 s
Wall time: 4min 15s


In [13]:
from test_utils import *
from NNCPD_Puchkin24 import *
import time
dir_path = "../data/mnist/"

files = get_files_list(dir_path)
files.sort()


CPS_NN = {}
RT_NN = []
ind_cps = 0
for afile in files:
    
    data = pd.read_csv(afile, index_col=False)
    label = data["Label"].values
    X = data.drop(columns=["Time", "Label"]).values
    X = np.nan_to_num(X)
    T = np.arange(len(X))

    # apply standard scaler
    X = StandardScaler().fit_transform(X)
      
    change_points = list(T[label == 1]) + [len(X)]


        


    # Apply the change point detection procedure
    # with the class of neural networks

    # Initialization
    st_nn = 0
    new_st_nn = 0

    # the threshold
    z_nn = 25

    # Initialization of the test statistic
    S_nn = np.empty(0)

    # Initialization of the list of detected change points
    change_points_nn = []


    # Initialization of the delays array and
    # the false alarms counter
    delays_nn = np.empty(0)
    current_change_point_ind = 0
    false_alarms_nn = 0

    start = time.time()
    while new_st_nn >= 0:
        
        # Run the procedure until the moment
        # it reports a change point occurrence
        new_S_nn, new_st_nn = compute_test_stat_nn(X[st_nn + 1:], z_nn, n_epochs=5)
        
        S_nn = np.append(S_nn, new_S_nn)
        
        st_nn += new_st_nn
        change_points_nn += [int(st_nn)]
        
        if(new_st_nn > 0):
            print('Detected change point (neural networks):', st_nn)
            
        if (st_nn < change_points[current_change_point_ind]):
            false_alarms_nn += 1
        
        else:
            delays_nn = np.append(delays_nn, np.array([st_nn - change_points[current_change_point_ind]]), axis=0)
            current_change_point_ind += 1
    end = time.time()
    print('Neural networks. Number of false alarms:', false_alarms_nn, '; average delay:', np.mean(delays_nn),\
        '±', np.std(delays_nn))
    RT_NN.append(end-start)
    CPS_NN[ind_cps] = change_points_nn
    ind_cps+=1  
    print(change_points)


Detected change point (neural networks): 199
Detected change point (neural networks): 382
Detected change point (neural networks): 559
Detected change point (neural networks): 742
Detected change point (neural networks): 923
Detected change point (neural networks): 1105
Detected change point (neural networks): 1285
Detected change point (neural networks): 1463
Detected change point (neural networks): 1640
Neural networks. Number of false alarms: 1 ; average delay: 21.666666666666668 ± 0.816496580927726
[178, 360, 537, 720, 901, 1083, 1264, 1443, 1617, 1797]
Detected change point (neural networks): 200
Detected change point (neural networks): 384
Detected change point (neural networks): 558
Detected change point (neural networks): 739
Detected change point (neural networks): 922
Detected change point (neural networks): 1107
Detected change point (neural networks): 1286
Detected change point (neural networks): 1465
Detected change point (neural networks): 1641
Neural networks. Number of 

In [20]:
from SWCPD import BaseDetector as SWDCP
from test_utils import *
import time
from sklearn.preprocessing import MinMaxScaler

In [22]:
CPS_NN

{0: [199, 382, 559, 742, 923, 1105, 1285, 1463, 1640, 1639],
 1: [200, 384, 558, 739, 922, 1107, 1286, 1465, 1641, 1640],
 2: [198, 383, 562, 740, 923, 1104, 1284, 1463, 1639, 1638],
 3: [200, 381, 561, 740, 923, 1108, 1287, 1467, 1637, 1636],
 4: [199, 381, 559, 740, 924, 1104, 1285, 1465, 1640, 1639],
 5: [199, 387, 559, 742, 924, 1105, 1286, 1466, 1644, 1643],
 6: [199, 384, 562, 740, 923, 1105, 1284, 1466, 1638, 1637],
 7: [198, 385, 561, 741, 922, 1104, 1284, 1464, 1643, 1642],
 8: [201, 381, 559, 741, 921, 1105, 1286, 1465, 1640, 1639],
 9: [199, 382, 559, 743, 923, 1107, 1286, 1463, 1641, 1640]}

In [41]:
RT_NN

[286.545711517334,
 285.43733310699463,
 284.4674618244171,
 283.05568861961365,
 284.5523386001587,
 285.2369034290314,
 284.5166492462158,
 283.97826313972473,
 283.68565368652344,
 281.702748298645]

In [27]:
detector = SWDCP(data=X.astype(np.float32),window_length=100,max_history=100,significance=0.075,use_cuda=False)

keys = ['AUC','FP','COV','DD']



METRICS_NN = {m: { k: [] for k in keys }  for m in [10,20,30,50]}

GT = [178, 360, 537, 720, 901, 1083, 1264, 1443, 1617]
for k,cps in CPS_NN.items():
    detector.change_points['loc'] = cps

    for m in [10,20,30,50]:
        f1, cov, auc_score, fp, delay = detector.evaluate(GT,m)
        METRICS_NN[m]['AUC'].append(auc_score)
        METRICS_NN[m]['COV'].append(cov)
        METRICS_NN[m]['FP'].append(fp)
        METRICS_NN[m]['DD'].append(delay[1])

for k in METRICS_NN.keys():
    print(k)
    print(f"Mean AUC {np.mean((METRICS_NN[k]['AUC']))} +- {np.std((METRICS_NN[k]['AUC']))}")
    print(f"Mean Cov {np.mean((METRICS_NN[k]['COV']))} +- {np.std((METRICS_NN[k]['COV']))}")
    print(f"Mean FP {np.mean((METRICS_NN[k]['FP']))} +- {np.min((METRICS_NN[k]['FP'])),np.max((METRICS_NN[k]['FP']))}")
    print(f"Mean FP {np.mean((METRICS_NN[k]['DD']))} +- {np.min((METRICS_NN[k]['DD'])),np.max((METRICS_NN[k]['DD']))}")


10
Mean AUC 0.09545454545454546 +- 0.0
Mean Cov 0.8023735259926532 +- 0.003887795357850402
Mean FP 10.0 +- (10, 10)
Mean FP 22.070000000000004 +- (21.4, 23.5)
20
Mean AUC 0.2481818181818182 +- 0.10629368329039134
Mean Cov 0.8023735259926532 +- 0.003887795357850402
Mean FP 8.3 +- (6, 10)
Mean FP 22.070000000000004 +- (21.4, 23.5)
30
Mean AUC 0.9545454545454547 +- 1.1102230246251565e-16
Mean Cov 0.8023735259926532 +- 0.003887795357850402
Mean FP 0.0 +- (0, 0)
Mean FP 22.070000000000004 +- (21.4, 23.5)
50
Mean AUC 0.9545454545454547 +- 1.1102230246251565e-16
Mean Cov 0.8023735259926532 +- 0.003887795357850402
Mean FP 0.0 +- (0, 0)
Mean FP 22.070000000000004 +- (21.4, 23.5)


In [28]:
dir_path = "../data/mnist/"


files = get_files_list(dir_path)
files.sort()

keys = ['AUC','FP','COV','DD']



METRICS = {m: { k: [] for k in keys }  for m in [10,20,30,50]}

print(METRICS)

RT = []

for afile in files:
    data = pd.read_csv(afile, index_col=False)

        # decompose data
    label = data["Label"].values
    X = data.drop(columns=["Time", "Label"]).values
    X = np.nan_to_num(X)
    #X = MinMaxScaler().fit_transform(X)
    T = np.arange(len(X))
    y_true = list(T[label == 1]) + [len(X)]
    print(y_true)

    detector = SWDCP(data=X.astype(np.float32),window_length=100,max_history=100,significance=0.075,use_cuda=False)
    start = time.time()
    detector.process_dataloader(p=2,n_theta=1000,seed=2025,verbose=False)
    end = time.time()
    RT.append(end-start)
    for m in [10,20,30,50]:
        f1, cov, auc_score, fp, delay = detector.evaluate(y_true[:-1],m)
        METRICS[m]['AUC'].append(auc_score)
        METRICS[m]['COV'].append(cov)
        METRICS[m]['FP'].append(fp)
        METRICS[m]['DD'].append(delay[1])


{10: {'AUC': [], 'FP': [], 'COV': [], 'DD': []}, 20: {'AUC': [], 'FP': [], 'COV': [], 'DD': []}, 30: {'AUC': [], 'FP': [], 'COV': [], 'DD': []}, 50: {'AUC': [], 'FP': [], 'COV': [], 'DD': []}}
[178, 360, 537, 720, 901, 1083, 1264, 1443, 1617, 1797]


100%|██████████| 1698/1698 [00:04<00:00, 388.52it/s]


[178, 360, 537, 720, 901, 1083, 1264, 1443, 1617, 1797]


100%|██████████| 1698/1698 [00:04<00:00, 390.36it/s]


[178, 360, 537, 720, 901, 1083, 1264, 1443, 1617, 1797]


100%|██████████| 1698/1698 [00:04<00:00, 405.76it/s]


[178, 360, 537, 720, 901, 1083, 1264, 1443, 1617, 1797]


100%|██████████| 1698/1698 [00:04<00:00, 401.09it/s]


[178, 360, 537, 720, 901, 1083, 1264, 1443, 1617, 1797]


100%|██████████| 1698/1698 [00:04<00:00, 394.02it/s]


[178, 360, 537, 720, 901, 1083, 1264, 1443, 1617, 1797]


100%|██████████| 1698/1698 [00:04<00:00, 399.97it/s]


[178, 360, 537, 720, 901, 1083, 1264, 1443, 1617, 1797]


100%|██████████| 1698/1698 [00:04<00:00, 416.87it/s]


[178, 360, 537, 720, 901, 1083, 1264, 1443, 1617, 1797]


100%|██████████| 1698/1698 [00:04<00:00, 408.06it/s]


[178, 360, 537, 720, 901, 1083, 1264, 1443, 1617, 1797]


100%|██████████| 1698/1698 [00:04<00:00, 406.52it/s]


[178, 360, 537, 720, 901, 1083, 1264, 1443, 1617, 1797]


100%|██████████| 1698/1698 [00:04<00:00, 403.17it/s]


In [37]:
np.mean(RT) ,np.std(RT)

(4.251901030540466, 0.1082743949244289)

In [29]:
for k in METRICS.keys():
    print(k)
    print(f"Mean AUC {np.mean((METRICS[k]['AUC']))} +- {np.std((METRICS[k]['AUC']))}")
    print(f"Mean Cov {np.mean((METRICS[k]['COV']))} +- {np.std((METRICS[k]['COV']))}")
    print(f"Mean FP {np.mean((METRICS[k]['FP']))} +- {np.min((METRICS[k]['FP'])),np.max((METRICS[k]['FP']))}")
    print(f"Mean FP {np.mean((METRICS[k]['DD']))} +- {np.min((METRICS[k]['DD'])),np.max((METRICS[k]['DD']))}")


10
Mean AUC 0.5122727272727272 +- 0.08234803183505623
Mean Cov 0.8960204422907649 +- 0.006585727638095648
Mean FP 5.1 +- (4, 6)
Mean FP 11.101111111111113 +- (9.333333333333334, 12.444444444444445)
20
Mean AUC 0.9663636363636364 +- 0.03858017304349625
Mean Cov 0.8960204422907649 +- 0.006585727638095648
Mean FP 0.2 +- (0, 1)
Mean FP 11.101111111111113 +- (9.333333333333334, 12.444444444444445)
30
Mean AUC 0.9863636363636363 +- 0.020829889522526526
Mean Cov 0.8960204422907649 +- 0.006585727638095648
Mean FP 0.0 +- (0, 0)
Mean FP 11.101111111111113 +- (9.333333333333334, 12.444444444444445)
50
Mean AUC 0.9863636363636363 +- 0.020829889522526526
Mean Cov 0.8960204422907649 +- 0.006585727638095648
Mean FP 0.0 +- (0, 0)
Mean FP 11.101111111111113 +- (9.333333333333334, 12.444444444444445)


|Method|AUC||FP||||
|---|---|---|---|---|---|---|
||$\tau=20$| $\tau=30$|$\tau=20$|$\tau=30$| **Cov**|**DD**|
|SWCDP|$0.97 \pm 0.04$|$0.99 \pm 0.02$|$0.2$ (0;1)|$0.0$ (0;0)|$0.90 \pm 0.01$|$11.1$ (9.3;12.4)|
|DeepClf [1]|$0.86 \pm 0.09$|$0.96 \pm 0.08$|$1.5$ (0;3)|$0.9$ (0;2)|$0.89 \pm 0.02$|$10.7$ (5.6;17)|
|DeepRulif [1]|$0.80 \pm 0.09$|$0.90 \pm 0.05$|$2.2$ (0;4)|$1.2$ (0;3)|$0.88 \pm 0.02$|$12.1$ (7.1;18)|
|DeepNN [2]|$0.24 \pm 0.11$|$0.96 \pm 0.01$|$8.3$ (6;10)|$0.0$ (0;0)|$0.80 \pm 0.02$|$22.1$ (21.4;23.5)|
|Rulif [3]|$0.72 \pm 0.09$|$0.84 \pm 0.1$|$1.6$ (0;3)|$0.4$ (0;1)|$0.66 \pm 0.02$|$11.5$ (6.8;14.3)|


# SUSY

In [10]:
%%time

dir_path = "../data/susy/"

models = [wrappers.RupturesBinseg(model='rbf', jump=10), 
          wrappers.RupturesPelt(model='rbf', jump=10), 
          wrappers.RupturesWindow(model='rbf', jump=10, width=100), 
          wrappers.OnlineCLF(lag_size=100, height=.1), 
          wrappers.OnlineRuLSIF(lag_size=100, height=.1), 
          wrappers.OrigRuLSIF(alpha=0.1, kernel_num=10, window_size=100, step=10, periods=1, height=0.1)]

for amodel in models:
    report = run_test_on_dataset(dir_path, amodel, downsample=10, margin=50, verbose=0, sigma=2)
    print(amodel)
    display(report)

,0,1
RI,0.983040,0.006288
F1,0.897502,0.098246


,0,1
RI,0.983205,0.005641
F1,0.919298,0.078012


,0,1
RI,0.973265,0.011171
F1,0.833694,0.102336


,0,1
RI,0.983744,0.004759
F1,0.989473,0.022191


,0,1
RI,0.976334,0.004885
F1,0.966666,0.053672


,0,1
RI,0.954352,0.017995
F1,0.762416,0.139424


CPU times: user 16.2 s, sys: 2.49 s, total: 18.6 s
Wall time: 17min 41s


# HTRU2

In [11]:
%%time

dir_path = "../data/htru2/"

models = [wrappers.RupturesBinseg(model='rbf', jump=10), 
          wrappers.RupturesPelt(model='rbf', jump=10), 
          wrappers.RupturesWindow(model='rbf', jump=10, width=100), 
          wrappers.OnlineCLF(lag_size=100, height=.1), 
          wrappers.OnlineRuLSIF(lag_size=100, height=.1), 
          wrappers.OrigRuLSIF(alpha=0.1, kernel_num=10, window_size=100, step=10, periods=1, height=0.1)]

for amodel in models:
    report = run_test_on_dataset(dir_path, amodel, downsample=10, margin=50, verbose=0, sigma=5)
    print(amodel)
    display(report)

,0,1
RI,0.984690,0.011205
F1,0.905222,0.114781


,0,1
RI,0.981880,0.010415
F1,0.893715,0.102571


,0,1
RI,0.967245,0.008168
F1,0.818190,0.091582


,0,1
RI,0.981505,0.005325
F1,0.978362,0.038303


,0,1
RI,0.973949,0.006867
F1,0.934444,0.056522


,0,1
RI,0.959620,0.009531
F1,0.849756,0.084128


CPU times: user 13.4 s, sys: 2.45 s, total: 15.9 s
Wall time: 17min 18s


# MAGIC

In [12]:
%%time

dir_path = "../data/magic/"

models = [wrappers.RupturesBinseg(model='rbf', jump=10), 
          wrappers.RupturesPelt(model='rbf', jump=10), 
          wrappers.RupturesWindow(model='rbf', jump=10, width=100), 
          wrappers.OnlineCLF(lag_size=100, height=.1), 
          wrappers.OnlineRuLSIF(lag_size=100, height=.1), 
          wrappers.OrigRuLSIF(alpha=0.1, kernel_num=10, window_size=100, step=10, periods=1, height=0.1)]

for amodel in models:
    report = run_test_on_dataset(dir_path, amodel, downsample=10, margin=50, verbose=0, sigma=2)
    print(amodel)
    display(report)

,0,1
RI,0.964965,0.013177
F1,0.677122,0.190681


,0,1
RI,0.971185,0.016209
F1,0.831365,0.096374


,0,1
RI,0.965390,0.008697
F1,0.769203,0.085897


,0,1
RI,0.963498,0.008331
F1,0.882078,0.077420


,0,1
RI,0.966791,0.012904
F1,0.869556,0.137267


,0,1
RI,0.850414,0.094979
F1,0.584186,0.166599


CPU times: user 13.8 s, sys: 2.3 s, total: 16.1 s
Wall time: 17min 40s


# Higgs

In [13]:
%%time

dir_path = "../data/higgs/"

models = [wrappers.RupturesBinseg(model='rbf', jump=10), 
          wrappers.RupturesPelt(model='rbf', jump=10), 
          wrappers.RupturesWindow(model='rbf', jump=10, width=200), 
          wrappers.OnlineCLF(lag_size=200, height=.1), 
          wrappers.OnlineRuLSIF(lag_size=200, height=.1), 
          wrappers.OrigRuLSIF(alpha=0.1, kernel_num=10, window_size=100, step=10, periods=1, height=0.1)]

for amodel in models:
    report = run_test_on_dataset(dir_path, amodel, downsample=5, margin=50, verbose=0)
    print(amodel)
    display(report)

,0,1
RI,0.959744,0.017775
F1,0.513627,0.167930


,0,1
RI,0.908315,0.010563
F1,0.175431,0.260583


,0,1
RI,0.954359,0.008388
F1,0.523818,0.141527


,0,1
RI,0.971241,0.011186
F1,0.757237,0.159999


,0,1
RI,0.974457,0.005138
F1,0.761871,0.105362


,0,1
RI,0.747915,0.233792
F1,0.232490,0.207845


CPU times: user 42.6 s, sys: 11.5 s, total: 54.1 s
Wall time: 1h 32min 59s


# EMG

In [14]:
%%time

dir_path = "../data/EMG_Physical_Action_Data_Set/"

models = [wrappers.RupturesBinseg(model='rbf', jump=10), 
          wrappers.RupturesPelt(model='rbf', jump=10), 
          wrappers.RupturesWindow(model='rbf', jump=10, width=100), 
          wrappers.OnlineCLF(lag_size=100, height=.1), 
          wrappers.OnlineRuLSIF(lag_size=100, height=.1), 
          wrappers.OrigRuLSIF(alpha=0.1, kernel_num=10, window_size=100, step=10, periods=1, height=0.1)]

for amodel in models:
    report = run_test_on_dataset(dir_path, amodel, downsample=50, margin=50, verbose=0)
    print(amodel)
    display(report)

,0,1
RI,0.969000,0.012791
F1,0.898488,0.029202


,0,1
RI,0.966179,0.010723
F1,0.886792,0.041223


,0,1
RI,0.974369,0.012075
F1,0.818434,0.107471


,0,1
RI,0.979054,0.009605
F1,0.972222,0.055556


,0,1
RI,0.980058,0.004711
F1,0.973684,0.030387


,0,1
RI,0.972983,0.018324
F1,0.954044,0.059547


CPU times: user 6.36 s, sys: 1.13 s, total: 7.49 s
Wall time: 5min 29s


# WISDM

In [15]:
%%time

dir_path = "../data/wisdm/"

models = [wrappers.RupturesBinseg(model='rbf', jump=10), 
          wrappers.RupturesPelt(model='rbf', jump=10), 
          wrappers.RupturesWindow(model='rbf', jump=10, width=100), 
          wrappers.OnlineCLF(lag_size=100, height=.1), 
          wrappers.OnlineRuLSIF(lag_size=100, height=.1), 
          wrappers.OrigRuLSIF(alpha=0.1, kernel_num=10, window_size=100, step=10, periods=1, height=0.1)]

for amodel in models:
    report = run_test_on_dataset(dir_path, amodel, downsample=20, margin=50, verbose=0)
    print(amodel)
    display(report)

,0,1
RI,0.990174,0.005106
F1,0.882233,0.118083


,0,1
RI,0.989973,0.005085
F1,0.864918,0.122434


,0,1
RI,0.990585,0.004573
F1,0.935753,0.061937


,0,1
RI,0.989948,0.004591
F1,0.958640,0.057770


,0,1
RI,0.98939,0.005578
F1,0.97344,0.040312


,0,1
RI,0.988019,0.003424
F1,0.940910,0.064357


CPU times: user 25.9 s, sys: 6.09 s, total: 31.9 s
Wall time: 24min 8s


# Kepler

In [16]:
%%time

dir_path = "../data/kepler/"

models = [wrappers.RupturesBinseg(model='rbf', jump=10), 
          wrappers.RupturesPelt(model='rbf', jump=10), 
          wrappers.RupturesWindow(model='rbf', jump=10, width=20), 
          wrappers.OnlineCLF(lag_size=20, height=1.1), 
          wrappers.OnlineRuLSIF(lag_size=20, height=1.1), 
          wrappers.OrigRuLSIF(alpha=0.1, kernel_num=10, window_size=20, step=10, periods=1, height=1.1)]

for amodel in models:
    report = run_test_on_dataset(dir_path, amodel, downsample=5, margin=20, verbose=0)
    print(amodel)
    display(report)

,0,1
RI,0.949039,0.023165
F1,0.596422,0.062363


,0,1
RI,0.990544,0.012027
F1,0.965490,0.043845


,0,1
RI,0.985066,0.009546
F1,0.881030,0.047975


,0,1
RI,0.997287,0.001685
F1,0.996621,0.009555


,0,1
RI,0.997307,0.002037
F1,0.971590,0.080353


,0,1
RI,0.889837,0.031741
F1,0.136065,0.041847


CPU times: user 1min 42s, sys: 29.9 s, total: 2min 12s
Wall time: 1h 26min 13s
